### ISSUE TITLE
hs.image, hs.drawing, and going forward

### ISSUE THREAD

**Issue by asmagill** [Hammerspoon/hammerspoon#425]
I want to create a method to composite images, which naturally falls within the hs.image domain, but since I'd like to be able to do things like place highlights on screen shots, etc, there will be some overlap with rectangles, etc. currently found in hs.drawing... this leads to what I see as three logical possible paths...
1. Modify hs.drawing, so it's rectangle/circle/text methods actually return hs.image objects, so hs.image can composite them without hassle.
2. Modify hs.image so it internally converts/copies hs.drawing objects to hs.image objects... this has the benefit that, until an hs.drawing is actually embedded/copied into an hs.image, it can be manipulated by the appropriate methods more accurately than simply scaling/modifying a bitmap might.
3. Copy the relevant methods to hs.image, so that they return hs.image objects, but leave hs.drawing as-is for simple one-offs.

I'm kind of leaning towards 1 or 2... I see 1 as easier, but 2 probably more flexible in the long run, as I also want to eventually create a hs.text (or something like it) module which gives a lot more control over text images than hs.drawing.text currently does... this way, hs.drawing provides simple drawing objects (maybe add bezier curves, etc.), hs.text provides textual objects, and hs.image handles image manipulation (maybe also move hs.drawing.setFrame,etc. and hs.drawing.image into hs.image so the modules are more internally consistent as to their purpose).

How much would break in terms of things people are working on right now, if these changes occurred (i.e. how much "backwards compatibility" should I try to ensure)?

Thoughts?

**Comment by cmsj** [comment]
I'm not opposed to this, but I am curious what the use case is for Hammerspoon marking up screenshots/etc. In general I think image composition could be very handy for making objects to feed into hs.drawing :)

In terms of those choices, I think either 2 or 3. My issue with 1 is that it moves hs.drawing objects away from their rendering primitives, which can be used at any time to modify the drawing object without having to rasterise and reload an image. I'm tempted to say 3, because these are all fairly trivial objects, but I think 2 is the ideal option, and possibly not too hard (we already have `window:snapshot()`)

As for backwards compatibility - I'm not sure that 2 or 3 would break anything? ATM there's no route from hs.drawing -> hs.image, only the reverse.

**Comment by asmagill** [comment]
A few basic cases where I see the current implementation as more difficult than it should be...

  * drawing something of multiple lines (triangle)... currently requires 3+ separate drawing objects, all of which have to be moved separately, if I want to move it around... maybe following the cursor for placement, for example

  * look at my font example (under my config on github, utils/_keys subdirectory) -- to display a list of font names is 100 drawing objects -- thank god it's stationary, but it still means 100 delete or hides to clear it. Granted, a fully functional hs.text (or similar) might make this a lot simpler, too.

  * when writing documentation, I'm often highlighting or marking up screen shots... automating some of this (when I can know pixel wise where things are) would be a huge benefit -- I'm also giving thought to extending hs.pasteboard to allow adding an image to the pasteboard  (and retrieving from?), but that's a little ways off at the moment.

As to my thoughts in reducing hs.drawing, I'm seeing it as more of a set of primitive graphical objects, and not the render engine... since hs.image includes the writing to file, it makes a certain amount of sense that it would render to the screen as well, but maybe not...perhaps making a hs.canvas or hs.primitives, which then feeds into hs.drawing or hs.image makes more sense? (hs.drawing.rectangle/etc would then be wrappers to hs.primitives which includes screen rendering, kind of like hs.hotkey.bind combines new and enable)

In thinking it over, I do want lines, rectangles, etc. to stay non-bitmaps up until hs.image actually does something to them -- composite so it can be passed to whatever rendering mechanism (hs.drawing, hs.menubar.icon, etc.), save to file, etc. ... it makes it easier to scale and position them, i think.


At any rate, these are just thoughts on the direction of the modules, not actual code in progress -- yet.  I want to make sure we're defining these in the most logical way for growth, not just the most expedient for "now".  "Expedient" has a way of becoming legacy...

**Comment by cmsj** [comment]
Fair enough :)

So given the intersection of our preferences is #2, is that the plan?

**Comment by lowne** [comment]
fwiw, I agree with the general principles outlined by @asmagill; this is a possible simplified roadmap:

1. deprecate everything in `hs.drawing`
2. to simplify, let's assume a new `hs.draw` module (alternatively we can still keep them in hs.drawing (maintaining backward compatibility in this case will be a bit ugly, but still doable); or have separate first-level hs.primitives and hs.text) with:
  - `.rectangle()`, etc constructors (all primitives)
  - stuff like `.imageFromASCII(...)` if it makes sense (and if they ever exists in a non-bitmap state)
  - `.text(...)` constructor (and long-needed features such as rich text, some basic layouting, etc.)
  - the usual `:setColor()` etc methods
  - for convenience, a `:rasterize(rect)` method that returns an `hs.image` (it'd call an hs.image.imageFromDrawing constructor)
3. everything in `hs.draw` stays 'vector'; hs.draw by itself can't output anything, the only way out is :rasterize
4. `hs.image` can take hs.draw stuff and is in charge of all rasterizing, e.g. 
  - `.imageFromDrawing(drawing,rect)` -- makes an image of the desired size
  - `:saveToFile([rect])` --optional parameter for post-raster resize
  - `:show([rect])` -- create the window etc., can use the constructor rect if omitted here
  - `.compose(...)`
5. implement backward compatibility layer for the current `hs.drawing`, by calling hs.draw and hs.image as appropriate (and keeping constructor rects and image objects saved locally for referencing in the legacy `hs.drawing:show()`, `:setFrame(f)` etc.)
6. eventually switch everything that uses hs.drawing or any sort of bitmap to use hs.image; but we should keep conveniences/shortcuts such as in the current hs.menubar.setIcon - the best way to achieve this is via a pseudo-polymorphic (yay type dance! :dancer: ) `hs.image.new` constructor, and just pass everything along

**Comment by cmsj** [comment]
@lowne one downside of that plan would be that you lose the ability to conveniently change a drawing after it has been displayed, you'd have to change the object, re-rasterise it and then display it.

I think there is still a substantial case to needs to be made, to justify making very deep changes to the way hs.image/hs.drawing work. I take @asmagill's point that it's currently not possible to draw arbitrary polygons/paths. We could fix that pretty easily by accepting tables of points that get mapped onto NSBezierPath, and that's just another subclass of HSDrawingView with the relevant options (although at the implementation layer, that's what ASCIImage does).

Even stuff like somehow making hs.drawing.text support rich text and/or layout options, is a really huge amount of work, and I'm currently very sceptical that it's worth the effort. I do want Hammerspoon to get steadily more capable, but I worry that sometimes we're being pulled towards architecture and features for their own sake, or for use cases that really should be solved in other ways.

So, let's see if we can map out some serious use cases that justify serious amounts of work :)

**Comment by asmagill** [comment]
I know it's long, but I'm trying to combine all of the above comments into something maybe a little more specific so we can define steps and targets...

Having thought some more, I can see a case for keeping the hs.drawing:show() method (which renders to the screen) as separate from hs.image -- hs.image deals (mostly) in bitmaps, things that can't easily change size, once they're "rendered" -- a bitmap based image file, output to a file, etc. (the fact that imageFromSystemName and *some* image files can scale well is due to the fact that OS X doesn't treat them as bitmaps *until they have to be*, and even then remembers the source in case it decides to redo it -- not because of anything inherent in hs.image itself)  hs.drawing keeps things in a more scalable format up until it is actually rendered on the screen, and even then, as long as it stays an NSView+NSWhatever at it's core, it can still usually be manipulated *on the screen*, but not with other mediums.

A case could also be made that hs.drawing:show(), setFrame(), etc. belong in one module, and the rest of the hs.drawing methods belong in another (like hs.primitives, hs.draw, whatever) and these then feed into hs.drawing (for screen rendering) or hs.image (for other rendering), because @cmsj is correct -- once something is converted to hs.image, it is more or less immutable, at least in a "clean" or "vectorized" way.  But this is more a question of organization than functionality, so other than keeping common code together and limiting the need to copy and paste into other module source files when we update/modify one of the other modules, I don't have a real strong opinion here. 

I'm thinking (without getting into which module "owns" these) we need:

* a way to group a bunch of drawing objects into 1 drawing object, creating an implicit transparent rectangle behind all of the included objects.
  * this transparent background's topLeft becomes the origin (0,0) from which member objects are offset
  * if you've saved the userdata for a member object, and modify it directly, it stays within the group
    * a change to position or level are relative to it's new origin.
    * show/hide to a member object directly affect whether or not it is shown/hidden when the grouped object is shown/hidden (i.e. if the group is hidden, but a member is sent show() (or hide()), the member object stays hidden, but *will* show (or remain hidden) when the group is next sent show().)
    * should there be a way to reference a member object via index, if you don't want to save a reference before grouping? e.g. member[x] where x is the order in which the objects are added to the group?
  * member objects have their base display level when rendered at that of the grouped object (i.e. the transparent background rectangle).  Attempts to change this directly via a reference from an outside reference to the member's userdata should be ignored (converted to relative to the rectangle?), unless it is relative to another object within the group.
  * moving/resizing the grouped object moves (and scales?) the member objects *relative to their position in their new origin rectangle*
  * setting the stroke/fill/color to this grouped object affects the rectangle background, not the member objects.
  * show/hide of the grouped object affects all members of the group, including the background
  * delete of the grouped object ungroups the member objects and deletes the implicit background... member objects should not be deleted, in case a reference to them still exists elsewhere -- allow __gc to handle them if/as necessary. 
  * do we need a way to "ungroup" the grouped object into a table of separate drawing objects, removing the implicit background at such time?

* additional primitive graphical objects
  * polygon
  * bezier curve
  * arc (i.e. partial circle -- bezier curve may be able to provide this without requiring a separate function)
  * others?

* text control (maybe as part of hs.drawing, maybe as separate module... ambivalent on this point)
  * I've not looked beyond the basics of NSFont and such, so I don't know for certain what we should/should not do, but at a minimum, we need better information about how it will be rendered (for calculating adjustments and other effects) within the defined "rectangle" of the text object, and possibly a way to adjust the properties which can be adjusted, such as:
    * actual max height and width of rendered text for the given font and size and style
    * baseline position relative to rectangle of object (to gauge/determine if text is going to render outside of it's given rectangle)
    * a way to determine the "frame" of a specific character (or character position) in the text so we know where to place other objects on top of it (for highlighting in a calendar, for example)
  * the grouping section above would make building complex styled text easier, maybe reducing the need for RTF support (I recognize RTF would still be easier for some cases, but see my next point).
  * if we leave out full RTF support here (and the above would make including it even more challenging), I've been wanting to take a look at webkit for including a way to render webpage or webpage snippits on the screen -- perhaps RTF belongs in this and not drawing.
    * this new "type" should be passable into hs.drawing and/or hs.image, though likely with some reductions of its original flexibility within its own scope... down the road a bit, I think. 

* other
  * hs.drawing (or whatever) passed into hs.image creates a new image.  This image *could* be passed back into hs.drawing, but this new object won't be modifiable as anything but an image -- you can't break it up, resizing scales a bitmap, etc.
  * the grouping *could* be implementable as a table in lua, and then treating the table as an object... however, if you look at the font example I reference in a previous comment, it does some of this, but it's also slow to render (you can see the items pop-up/update one at a time) -- I think something which works on them strictly form the C side as a group, with possible optimizations, will be noticeably faster and make it easier to provide a "buffered  update" experience.

Thoughts?

**Comment by lowne** [comment]
Good point, drawings still need their own :show() - I bet the implementation is quite different anyway so DRY isn't a concern. 
What I wanted to emphasise with my simplified roadmap is that even with a tabula-rasa approach there are ways to maintain backwards compatibility (I even *enjoy* ugly glue code in Lua).

Architecture-wise, I agree we're nowhere near the point where a Big Rewrite of drawing-related stuff is warranted; the current image/drawing perceptual overlap can be a little confusing but surely not a major issue.

Functionality-wise, I think there's a case for improved text; I'm not talking about a Pages.app-level layout engine (in capabilities nor performance), but, as an example, let's take hs.expose (#436); I'd like to:
- overlay titles on top of window thumbnails; this requires something like #394, or a hardcoded fixed width font :cry:
- ditch the precoded hints in favour of fuzzy matching; this would need the ability to highlight the typed letters on app names in a different colour for visual feedback
In short, I'd just like #394 :) - it would allow a minimal layout engine to be built (relatively quickly) in Lua

Similarly, with arbitrary paths + grouping/compositing most everything can be then done in Lua; performance will always be lacklustre, but I'll happily pay that price in exchange of much faster coding (alternatively, LuaJIT, but that's a whole 'nother can of worms - ditching 5.3, integers etc.; it might be too late - :disappointed: #383)

**Comment by cmsj** [comment]
Without wanting to completely re-read this thread, I want to guess that the various improvements to `hs.image` over time, and now the introduction of `hs.canvas` mean I can close this issue? I'm going to do so, and anyone who disagrees is very welcome to re-open it :)
